processing/src/main/java/io/druid/query/groupby/epinephelinae/LimitedBufferGrouper.java

```diff
@@ -0,0 +1,627 @@
+  {
+    if (sortHasNonGroupingFields) {
+      // re-sort the heap in place, it's also an array of offsets in the buffer
+      return makeDefaultOrderingIterator(offsetHeap.getHeapSize());
+    } else {
+      return makeHeapIterator();
+    }
+  }
+
+  @Override
+  public void close()
+  {
+    for (BufferAggregator aggregator : aggregators) {
+      try {
+        aggregator.close();
+      }
+      catch (Exception e) {
+        log.warn(e, "Could not close aggregator, skipping.", aggregator);
+      }
+    }
+  }
+
+  public int getGrowthCount()
+  {
+    return growthCount;
+  }
+
+  public int getSize()
+  {
+    return size;
+  }
+
+  public int getBuckets()
+  {
+    return buckets;
+  }
+
+  public int getLimit()
+  {
+    return limit;
+  }
+
+  public int getMaxSize()
+  {
+    return maxSize;
+  }
+
+  private Iterator<Grouper.Entry<KeyType>> makeDefaultOrderingIterator(final int size)
+  {
+    final List<Integer> wrappedOffsets = new AbstractList<Integer>()
+    {
+      @Override
+      public Integer get(int index)
+      {
+        return buffer.getInt(tableArenaSize + index * Ints.BYTES);
+      }
+
+      @Override
+      public Integer set(int index, Integer element)
+      {
+        final Integer oldValue = get(index);
+        buffer.putInt(tableArenaSize + index * Ints.BYTES, element);
+        return oldValue;
+      }
+
+      @Override
+      public int size()
+      {
+        return size;
+      }
+    };
+
+    final Grouper.KeyComparator comparator = keySerde.bufferComparator();
+
+    // Sort offsets in-place.
+    Collections.sort(
+        wrappedOffsets,
+        new Comparator<Integer>()
+        {
+          @Override
+          public int compare(Integer lhs, Integer rhs)
+          {
+            return comparator.compare(
+                tableBuffer,
+                tableBuffer,
+                lhs + HASH_SIZE,
+                rhs + HASH_SIZE
+            );
+          }
+        }
+    );
+
+    return new Iterator<Grouper.Entry<KeyType>>()
+    {
+      int curr = 0;
+
+      @Override
+      public boolean hasNext()
+      {
+        return curr < size;
+      }
+
+      @Override
+      public Grouper.Entry<KeyType> next()
+      {
+        return bucketEntryForOffset(wrappedOffsets.get(curr++));
+      }
+
+      @Override
+      public void remove()
+      {
+        throw new UnsupportedOperationException();
+      }
+    };
+  }
+
+  private Iterator<Grouper.Entry<KeyType>> makeHeapIterator()
+  {
+    final int initialHeapSize = offsetHeap.getHeapSize();
+    return new Iterator<Grouper.Entry<KeyType>>()
+    {
+      int curr = 0;
+
+      @Override
+      public boolean hasNext()
+      {
+        return curr < initialHeapSize;
+      }
+
+      @Override
+      public Grouper.Entry<KeyType> next()
+      {
+        if (curr >= initialHeapSize) {
+          throw new NoSuchElementException();
+        }
+        final int offset = offsetHeap.removeMin();
```

Comment:
Hm, not sure if there's a way to get a sorted iterator from the heap non-destructively, an alternative would be build a list and sort it, but that introduces sorting overhead as well, they should both be N * logN operations on average